### ISSUE TITLE
[Workflow] Add a validator to check / enforce state change restrictions

### ISSUE THREAD

**Issue by dkarlovi** [symfony/symfony#23863]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | yes
| Symfony version  | latest

If you have an entity with a state machine attached, you can still just set any state as current and save if you're not using `apply()`.

It would make sense to add a validator specifically enforcing the attached state machine, it would:

1. figure out if the state being set is actually a valid state at all
2. figure out if the state being set is a valid state **from current state**
3. (this is debateable) try to apply the state so event listeners are able to block the transition?

It would allow for more easier integration of the component into common workflows, IMO.

**Edit:** if 3) is considered, listeners should probably be able to provide an explanation why they've blocked the transition (`ConstraintViolationInterface` or similar?)

WDYT?

**Comment by Nyholm** [comment]
Can you specify what you mean with "common workflows"?

> If you have an entity with a state machine attached, you can still just set any state as current and save if you're not using apply().

Sure, as you still can edit your database with SQL. ;)

**Comment by dkarlovi** [comment]
> Sure, as you still can edit your database with SQL. ;)

Touche. :)

> Can you specify what you mean with "common workflows"?

Sure.

In a lot of situations, developers will not manage their entities on their own but instead use some sort of abstraction (say, Sylius, Sonata, REST Bundle or in my case API Platform). Those platforms will, however, allow you to define both your entities and their serialization, validation as part of their common integration / usage path, it's pretty much expected in order to use them.

Having a validator here would allow for hooking workflows / state machines into the existing systems easily, having all the benefits like using their custom validation serialization (again, API platform).

In my use case, I have a entity `Image` which has a state machine attached. One of the transition paths is for the end user to request a new Image created which will be delivered directly to cloud storage (quite a common pattern), so you have:

```yaml
image:
- created
- upload_pending
- uploaded
```

User creates the image entity and it gets moved to `upload_pending` state at which point the user is assigned an upload URL to Azure Blob Storage. Once the upload is complete, user's client sends a `state: uploaded` back at which point we should change the state.

State should only be changed if it's a valid next state (which includes validating the image is actually uploaded, among other things). I have this pretty much completed, but feel like it would be much nicer to actually be able to use the Workflow component here directly as it's the source of truth here.

**Comment by lyrixx** [comment]
Hello.

> It would make sense to add a validator specifically enforcing the attached state machine, it would:

Just to be sure, you would like an integration of the workflow component in to validator component.

If yes, It makes sens to me. But only for the `can` method. The apply method can not be use in the validator component as the validator **should not** modify the state of the validated POPO.

How do you want to configure it?

```

class Image
{
    /**
     * @AssertWorkflow(
     *  message="This transition is not enabled",
     *  workflow="image"
     * )
     */
    public $marking
}
```

Is that ok for you ?

Do you want to create the PR?

**Comment by lyrixx** [comment]
Ouch. I write my previous message too fast.

Actually the previous configuration I wrote is totally wrong ! It can not work.

What we should test is the transition name, not a marking. So:

* how could you configure that?
* what does your client send ? (the new marking? (this is wrong), the transition name?)
* how do you map the the value send by your client to your entity?

---

I already used API Platform and IMHO the best way to handle this use case is by creating 2 API endpoint.

* The first one is the classic PUT (or PATCH) method. This one should not modify the `marking` property.
* The second one is a dedicated workflow endpoint. Something like `/image/{id}/transition/{transition}`.

I think it's better because the CRUD management is really different from the Workflow management.

Finally, If you are using a DTO, you can mixed theses 2 endpoints together, but I don't think it's a good idea.

**Comment by fmata** [comment]
Just my 2 cents but in one project using API Platform and symfony/workflow, I provide 2 endpoints as suggested by @lyrixx : 

- /workflows/{resourceName}/{resourceId}
Sends enabled transition for the current state of the resourceName entity #resourceId workflow. Not REST but very useful to have only one endpoint to handle all resources :)

- /images/{id]/transition/{transition}
Applies a transition after checks

```
public function transitionAction(Image $image, $transition)
    {
        $workflow = $this->get('state_machine.image_lifecycle');

        if (!$workflow->can($image, $transition)) {
            $enableTransitions = $workflow->getEnabledTransitions($image);

            if (!$enableTransitions) {
                throw new ValidatorException(sprintf("Image has a final status '%s' : no transition can be applied.", $image->getStatus()));
            }

            $availablesTransitions = [];
            foreach ($enableTransitions as $enableTransition) {
                $availablesTransitions[] = $enableTransition->getName();
            }

            throw new ValidatorException(sprintf("Image with the status '%s' cannot be '%s' : you can only apply '%s' transition(s)", $image->getStatus(), $transition, implode(', ', $availablesTransitions)));
        }

        $workflow->apply($image, $transition);

        return $image;
    }
```

So a native validator could be very useful.

About the first question, I have no good idea to handle the case to restrain the marking to be called publicly...

**Comment by dkarlovi** [comment]
> I have no good idea to handle the case to restrain the marking to be called publicly

I currently use a serializer group `image_update_state_uploaded` on the marking and set the group dynamically for the request only if appropriate. It makes the property invisible for update if not by proper user and in proper current state.

**Comment by dkarlovi** [comment]
@lyrixx 

> how could you configure that?

The example you've given seems as a OK first draft, to be honest.

> what does your client send ? (the new marking? (this is wrong), the transition name?)

I think the client should send the new actual state, so sending a transition name is basically a RPC here, State seems more "proper", but not crucial. This might be even more important [if API Platform implements JSON Patch support.](https://github.com/api-platform/core/issues/759). 

> how do you map the the value send by your client to your entity?

Directly, but as I said, it's made read-only by the serializer, only exception is this exact transition (everything else is done via ESB - RabbitMQ + Enqueue bundle + async event handling).

**Comment by lyrixx** [comment]
@dkarlovi 

> I think the client should send the new actual state, so sending a transition name is basically a RPC here, State seems more "proper", but not crucial. This might be even more important if API Platform implements JSON Patch support..

This is really wrong. The client SHOULD NOT send the new state. If you do that you bypass the the workflow component. If you need some arguments to convince you:

* You will duplicate the logic between the server and the client (even if the logic is really small)
* You can not guarante the marking of an object as you modify directly this marking.
* The marking of an object could be stored outside this one
* Is not really possible to validate if an object can be in state (X, Y, Z). The only way to do that is to test all combinaison of transition and ensure this marking is possible. But, even, This does not valide if the object where in state (A, B) before this.

@fmata :

> So a native validator could be very useful.

How would you use configure and use this validator?

**Comment by fmata** [comment]
@dkarlovi My first implementation was the same as your but with sending the status you loose the semantic of the change. A place could have many froms so you must add logic to your API to double check the right transition to apply. But I think it's not the right way to use symfony/workflow and its events (including guard related).

Or maybe the component is overkill in your use case.

**Comment by dkarlovi** [comment]
@lyrixx @fmata thanks for your comments, it's interesting to get another insight and other solutions for a specific problem even though we're somewhat distracting from the core of the issue. I'll address it real quick:

> The client SHOULD NOT send the new state. If you do that you bypass the the workflow component.

I disagree here. If we assume there's such a thing as a workflow validator (the core topic of this issue), your "state" is a property of your entity. So, it's not different than an "email" or a "title", it's a property which keeps track of something. It also requires the input to be valid, just the validation of that validity is more elaborate than e-mail regex or string length.

If we assume there's NO explicit state machine behind the state change, you'd send a state update exactly like that - update the state just as you'd update a title. Introducing the SM means y If you change from a state machine to something else, you'dou now send the transition instead. This means sending the transition is leaking your app's implementation over to the client.

> You can not guarante the marking of an object as you modify directly this marking.

That's the point behind the validator, **you're sending a request** to change the state.

> maybe the component is overkill in your use case.

I don't think so, [it uses the SM to figure out what's the next thing needed to be done and uses async event listeners over a message queue](https://github.com/php-enqueue/enqueue-dev/issues/68#issue-226194445), converting Doctrine lifecycle events to trigger the next message and a small wrapper around the Workflow component to figure out what to to next. Basically, the web app triggers the first Doctrine save and workers message each other about it until the state machine is in full stop, I'm even happy with the simplicity of the code which does all of this.

**Comment by fmata** [comment]
@lyrixx I can't see how to provide this validator... we need to know what change we want to do on the object but to achieve this we have to set the status as @dkarlovi already does, but IMHO it's not the right way.

So now I'm not sure if it's really possible to create a validator.

Maybe we can imagine an improvement to symfony/workflow to allow the marking store to call a private callable, using reflection. Keep the current behavior allowed and let those who want more constraint use a private setter.

**Comment by dkarlovi** [comment]
> it's not the right way

I don't think this should be an obstacle: if we have the current state and desired next state, we can identify the transition, if we have the current state and desired transition, we can identify next state. Functionally it should be equivalent and we shouldn't focus on the pedantics here too much. :)

Topic at hand is the Workflow validator and how to reliably confirm the transition can take place, hopefully with a `ConstraintViolation` explanation why not. Applying the transition is not a problem (if it doesn't fail), integrating it into the whole dispatch / validate / error / serialize is.

**Comment by lyrixx** [comment]
Hi @dkarlovi 

>> The client SHOULD NOT send the new state. If you do that you bypass the the workflow component.
> I disagree here.

Actually, the workflow component is build on top of the [petri net specification](https://en.wikipedia.org/wiki/Petri_net). So it should not be possible to set the marking directly in your object. Thus you can implement a validator in your own project but this is not something that can be integrated in Symfony because it's against the petri net spec.

I'm sorry and I'm going to close this issue.

**Comment by dkarlovi** [comment]
Hi @lyrixx fair enough. It's not worth doing it even with the input being a transition?

**Comment by lyrixx** [comment]
> It's not worth doing it even with the input being a transition?

Yes we could do something. But now I can not imagine how it could be configured / used.

Feel free to open a new issue with this description If you have something in mind ;)

**Comment by dkarlovi** [comment]
I'm doing a sample implementation in my app currently, will open a PR if it gets somewhere.

**Comment by dkarlovi** [comment]
@lyrixx working on this and have a rough understanding how things would fit together.

Do you think the `GuardEvent` could be extended to allow for a `ConstraintViolation` setter? This would allow listeners to explain why they've rejected the transition and could be directly used by the validator.

**Comment by lyrixx** [comment]
I'm not sure to fully understand. A transition can be blocked at several place:

* because there is not enough token in the previous places
* because a guard blocked it.

So IMHO, the first step is to just say (No possible / possible). We will see how we can enhance it after that.
Finally, In the symfony>security>voter: You don't know which voter declined your request. We lived with that for almost 6 years like that. So I'm fine with a basic implementation for now.

But I still don't see how this validator will be used, so I suggest you to not waste your time and a perfect implementation. IMHO we should discuss before.

**Comment by dkarlovi** [comment]
@lyrixx you're right, forgot about voters. :( My idea was to provide blocker the ability to explain why and use that explanation in the validator violation message.

Anyway, I have a PoC functional, it works like this (from perspective of API Platform flow and client setting the new state directly, as described above):
1. serialization and validation group on marking set to `image_update_state_uploaded`
2. `preDeserialize` listener notes the current state
3. serialization context builder will ask if the user is allowed to update state, voter allows it (if not, no state change)
4. `@serializer` deserializes
5. `postDeserialize` listener notes if there's a change from 2). If so, it resets marking to old state and sets new state to constraint instance (fetched from `@validator` metadata).
6. `@validator` validates, using the new Workflow validator (three conditions: a new state set at all, new state is valid from here, new state can be applied - not guarded)
7. `preWrite` listener applies the new state using the state machine
8. write

It's not trivial, but it does work and makes the client's interface to it much simpler IMO.

**Comment by lyrixx** [comment]
Thanks for posting your research here.
